pkg/storage/ingest/reader.go

```diff
@@ -94,20 +97,39 @@ func (r *PartitionReader) start(ctx context.Context) (returnErr error) {
 		}
 	}()
 
-	startFromOffset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
-	if err != nil {
-		return err
+	var startOffset int64
+	var err error
+
+	// Find the offset from which we should start consuming.
+	switch r.kafkaCfg.ConsumeFromPositionAtStartup {
+	case consumeFromStart:
+		startOffset = kafkaStartOffset
+		level.Info(r.logger).Log("msg", "starting consumption from partition start", "consumer_group", r.consumerGroup)
+
+	case consumeFromEnd:
+		startOffset = kafkaEndOffset
+		level.Warn(r.logger).Log("msg", "starting consumption from partition end (may cause data loss)", "consumer_group", r.consumerGroup)
+
+	default:
+		if offset, exists, err := r.fetchLastCommittedOffsetWithRetries(ctx); err != nil {
+			return err
+		} else if exists {
+			level.Info(r.logger).Log("msg", "starting consumption from last committed offset", "offset", offset, "consumer_group", r.consumerGroup)
+			startOffset = offset
+		} else {
+			level.Info(r.logger).Log("msg", "starting consumption from start because no committed offset has been found", "consumer_group", r.consumerGroup)
+			startOffset = kafkaStartOffset
+		}
 	}
-	level.Info(r.logger).Log("msg", "resuming consumption from offset", "offset", startFromOffset, "consumer_group", r.consumerGroup)
 
 	// Initialise the last consumed offset only if we've got a real offset from the consumer group.
 	// If we got a special offset (e.g. kafkaStartOffset) we want to keep the last consumed offset uninitialized,
 	// and it will be updated as soon as we consume the first record.
-	if startFromOffset >= 0 {
-		r.consumedOffsetWatcher.Notify(startFromOffset - 1)
+	if startOffset >= 0 {
+		r.consumedOffsetWatcher.Notify(startOffset - 1)
```

Comment:
If startOffset == 0, we will notify with -1. Is that OK? (Not new in this PR)

Reply:
"Yes" for now. It's still a mess how we handle offsets. One of my next PRs will try to cleanup a bit the logic, so that we can safely remove this `- 1`.